src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,431 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Net
+{
+    // Warning: Mutable struct!
+    // The purpose of this struct is to simplify buffer management in cases where the size of the buffer may grow large (e.g. >64K),
+    // thus making it worthwhile to add the overhead involved in managing multiple individual array allocations.
+    // Like ArrayBuffer, this manages a sliding buffer where bytes can be added at the end and removed at the beginning.
+    // Unlike ArrayBuffer, the buffer itself is managed using 16K blocks which are added/removed to the block list as necessary.
+
+    // [ActiveBuffer] contains the current buffer contents; these bytes will be preserved on any call to TryEnsureAvailableBytesUpToLimit.
+    // [AvailableBuffer] contains the available bytes past the end of the current content,
+    // and can be written to in order to add data to the end of the buffer.
+    // Commit(byteCount) will extend the ActiveBuffer by [byteCount] bytes into the AvailableBuffer.
+    // Discard(byteCount) will discard [byteCount] bytes as the beginning of the ActiveBuffer.
+    // TryEnsureAvailableBytesUpToLimit will grow the buffer if necessary; *however*, this may invalidate
+    // old values of [ActiveBuffer] and [AvailableBuffer], so they must be retrieved again.
+
+    internal struct MultiArrayBuffer : IDisposable
+    {
+        private byte[]?[]? _blocks;
+        private uint _allocatedEnd;
+        private uint _activeStart;
+        private uint _availableStart;
+
+        // Invariants:
+        // 0 <= _activeStart <= _availableStart <= total buffer size (i.e. _blockCount * BlockSize)
+
+        private const int BlockSize = 16 * 1024;
+
+        public MultiArrayBuffer(int initialBufferSize)
+        {
+            // [initialBufferSize] is ignored for now;
+            // I kept it because some callers are passing useful info here that we might want to act on in the future.
+            Debug.Assert(initialBufferSize >= 0);
+
+            _blocks = null;
+            _allocatedEnd = 0;
+            _activeStart = 0;
+            _availableStart = 0;
+        }
```

Comment:
Nit: it'd be a bit more efficient to do:
```C#
public MultiArrayBuffer(int initialBufferSize) : this()
{
    // [initialBufferSize] is ignored for now. Some callers are passing useful info here that we might want to act on in the future.
    Debug.Assert(initialBufferSize >= 0);
}
```

Reply:
Fixed